### ISSUE TITLE
use `schematics` instead of `volutuous`

### ISSUE THREAD

**Issue by riccardomurri** [gc3-uzh-ch/elasticluster#145]
Schematics is another Python library for data modelling and validation:

http://schematics.readthedocs.org/en/latest/index.html#

IMHO the advantages over volutuous are:

- much better error reporting
- should allow nested data models (whereas volutuous only checks the *values* of a Python `dict` instance)

**Comment by nicolasbaer** [comment]
You are right, that volutuous is not the best solution here. Creating a configuration file is an error-prone process and the first time you try elasticluster you'll most likely get it wrong. An improved error reporting should be the main concern when we change the configuration file validation. Currently we have a too complex configuration validation in place that does not even highlight the errors correctly. The overall process to validate the configuration file is: (i) create a configuration for each cluster and (ii) validate the configuration properties for each cluster. The first step is necessary to infer the properties based on inheritance (e.g. a node section can inherit properties from the cluster section) and to follow/denormalize the relations to other configuration sections. The second step then validates the configuration of each cluster to check whether the necessary fields are provided and correct. Now, the second step could be done with schematics. But if we validate the already parsed configuration, we will not be able to clearly specify in which section the error actually lies. Also we have to look into the nested models and how the error message looks like. If we have an error in `cluster.cloud.ec2_url` and schematics only reports an error in `ec2_url` we won't be happy :)

I'll simulate different error scenarios in the configuration file and report the exceptions here, so we can decide where to start improving. First, we should find out which of the above described steps produces bad error messages.

**Comment by riccardomurri** [comment]
Fixed in commit 94859c4a2d85c368c59ead73a2585dd65b1ff6a6
